### ISSUE TITLE
Implement gettext extraction within Timber

### ISSUE THREAD

**Issue by drzraf** [timber/timber#1754]
### Expected behavior
`wp-cli timber twig-pot`
Generate a Gettext pot file for the Timber-back theme.

### Actual behavior
POT for Twig is restricted to Poedit2 without any practical alternative.
That means it's unsuitable for collaborative i18n platforms which usually rely on continuous/automated POT upload/updates.

### But Extracting translatable strings needs the combined knowledge of several things:
1. How to parse a Twig template to extract specific node.
1. Know all the custom function/filters and register them
1. Where are the templates files located
1. How are called gettext function wrappers (`__;_e;_n;_x;_ex;...`)

### So which software can do that:
* Point 1 Can be done by any codebase relying on Twig parser + a couple of calls to `loadTemplate()`
* Point 2: A part can only be done by Timber (which contains several additional extensions). But a theme may declare other ones. So `functions.php` needs to be loaded too.
* Point 3 could be specified by hand/parameters to external tool, but is already known by Timber
* Point 4 could be specified by hand/parameters to external tool, but are already known by Timber

### Narrowing:
Point 2 exclude defacto external tools like `Twig-Gettext-Extractor` and imply WP bootstrap (`wp-cli`).
`Twig-Gettext-Extractor` compiles one user-provided Twig templates and run `xgettext()` on resulting PHP code. This is suboptimal and trying to give it the knowledge of WordPress Twig customizations not doable.
Moreover, passing through the `xgettext` `system()` call is probably not needed.
Twig parser is (_probably_) able to provide us every `callable` node whose `name` attribute is `_x` or `__`, ...
Provided such a list it should be fairly easy to generate to corresponding PO file without the need for `xgettext()`. See for example [how wp-cli i18n](https://github.com/wp-cli/i18n-command/blob/master/src/PhpCodeExtractor.php#L80) is creating [new Translation](https://github.com/oscarotero/Gettext/blob/master/src/Translation.php) from code.

### Conclusion
It's either the responsibility of `Timber` or `wp-cli i18n` to provide the feature (or a combination of both: `wp-cli` for bootstrapping, `wp-cli i18n` for Po-file format and `Timber` for Twig-related)

### Blocked by
- [x] [Avoid overly nested function call](https://github.com/timber/timber/pull/1753)
- [x] [AST-based php-gettext Twig parser (supporting the `Timber` loader)](https://github.com/php-gettext/Twig-Scanner)
- [x] Public release of php-gettext containing this scanner (v5)
- [ ] [wp-cli i18n command update php-gettext to v5](https://github.com/wp-cli/i18n-command/pull/217) : **On hold** because PHP < 7.2 is supported by WP (and consequently by wp-cli) but not by [php-gettext](https://github.com/php-gettext/Gettext/issues/270)
- [ ] new release of wp-cli + wp-cli i18n command

### References
- https://github.com/umpirsky/Twig-Gettext-Extractor/pull/60
Tagging @schlessera / @swissspidy

**Comment by gchtr** [comment]
Hey @drzraf 

Cool stuff 😎! I also thought that we need this for Timber when I saw that `i18n` was released for WP CLI. This would be really great to have 🙌!

I saw that you already dug deep in the linked issue. Of course it’s not the responsibility of the authors of that command to provide the functionality for Twig, but if we could hook into that functionality to make it work with Timber, that would be great!

If you need to make changes to Timber to make this work, pull requests would be very welcome! I didn’t look at it in detail myself, because you seem to have thought this through quite thoroughly already, but if we can help you in any way: tell us!

**Comment by drzraf** [comment]
Good to know that you're open to this. I'll try to provide a PR with Timber-specific bits.

IIRC, the issue I experienced was to easily get Timber Git-based installation: I couldn't get Timber to run:
- from Git (github)
- with all the composer dependencies
- while being activated as a wordpress's plugin (which is needed to load wp-cli commands).
- (with bedrock)
See also https://github.com/timber/timber/pull/1756
but I'll try again anyway.

**Comment by drzraf** [comment]
I just provided an initial implementation for Timber Twig gettext extraction at https://github.com/oscarotero/Gettext/.

I can now tell you something that needs fixing on Timber side:
The issue is that Timber is not usable when it's `composer require`d and autoloaded by another package.

When autoloaded, outside WordPress context (of before WordPress bootstraps), Timber constructor forcefully calls `init()` which makes use of `add_action()`. It's not even `\add_action()` and since neither `\add_action()` nor `\Timber\add_action()` does not exist, this fails.
=> See `ToDo` n°3 in [the PR](https://github.com/oscarotero/Gettext/pull/195).
This issue is probably similar to https://github.com/MINDKomm/timmy/issues/10

**Comment by gchtr** [comment]
Thanks for digging deep into this!

Hm... is the problem actually with `Timber\Timber::__construct()` or `Timber\Twig::__construct()`? In your pull request, I see that you’re creating an instance of `Timber\Twig`, which calls `add_action()`. It’s not the best idea to call WordPress from a class constructor, exactly because of the problem that you run into. But as I understand it, in `Timber\Timber::__construct()`, there’s already a check for whether WordPress is loaded, before `init()` is called:

https://github.com/timber/timber/blob/42cd28149e1c7c9894bc9eb2cff34e533b2c8553/lib/Timber.php#L51-L61

Would it help if we added this check to `Timber\Twig::__construct()` as well?

**Comment by rubas** [comment]
@drzraf Thank you for all the work. Were you able to solve the issue in the last 24 month? Looks like you got stuck on all ends :(

**Comment by drzraf** [comment]
https://github.com/php-gettext/Gettext/pull/195  would need to be considered by @oscarotero.
Once this done and part of a release, maybe reopen https://github.com/wp-cli/i18n-command/pull/59.

Anyway since that time, some PR have been merged to improve inheritability so that it's now possible to copy/paste the above PR files locally and bind them together by:
- extending `WP_CLI\I18n\PhpFunctionsScanner` (https://github.com/php-gettext/Gettext/pull/195 )
- extending `Gettext\Extractors\Twig`  (see https://github.com/wp-cli/i18n-command/issues/52#issuecomment-407493455)
- create a dedicated command (see https://github.com/wp-cli/i18n-command/issues/52#issuecomment-409281487)
- and a class extending `WP_CLI\I18n\MakePotCommand` to bridge all of this together.

**Comment by rubas** [comment]
Thanks. 

Lets see, if we can get your [PR](https://github.com/php-gettext/Gettext/pull/195) merged into Gettext. 
- @drzraf: Would you consider renaming PR to something like `Feedback wanted: ..` instead of `WIP...`. 

I guess this is the reason, why you haven't got any reaction yet.

---
A native `WP_CLI/i18n-command` integration would be perfect, but it is also something we can live without it.

**Comment by schlessera** [comment]
I'll just drop a series of related PRs to the `gettext` library in here that I managed to get merged _after_ our initial discussions and that might help (or not):

- https://github.com/php-gettext/Gettext/pull/232
- https://github.com/php-gettext/Gettext/pull/234
- https://github.com/php-gettext/Gettext/pull/235

**Comment by drzraf** [comment]
@gchtr : The ToDo n°3 in https://github.com/oscarotero/Gettext/pull/195 is about this code:
https://github.com/timber/timber/blob/11ef08aa9bde393bef6013bba224880a3d40533b/lib/Twig.php#L28-L31

`lib/Twig.php` actually fulfill two very different tasks:
1. One is tightly WP related: Binding PHP code (probably WP-specific) to a Twig function
2. The other, which is implicit but of broader interest, is **declaring** these Twig function **names**.

When extracting translations we may not be interested about what these custom Twig functions actually do.
But to parse templates, we still want to know which function names are declared.

As such, it's annoying that `lib/Twig.php` constructor forcefully calls WP-centric `add_action()` because we may want to use `add_timber_functions()` outside of a WordPress environment.

**Comment by gchtr** [comment]
@drzraf Thanks for taking this up again. Yes, we should change that. I’ve long changed to not calling hooks in constructors in my own projects. It’s something I stumble over sometimes in Timber and think: In the future this will become a problem. And here we are, it’s time to improve this.

I’m glad you start working on this again and you seem to have good overview over what’s needed. For your specific needs with gettext extraction, is it enough if we call hooks separately.

```php
<?php

class Twig {
	/**
	 * Inits hooks.
	 */
	public static function init() {
		add_action( 'timber/twig/filters', [ __CLASS__, 'add_timber_filters' ] );
		add_action( 'timber/twig/functions', [ __CLASS__, 'add_timber_functions' ] );
		add_action( 'timber/twig/escapers', [ __CLASS__, 'add_timber_escapers' ] );
	}
}
```

Or do we need something else to make the parser know which function names are declared? Because I’m not sure what you mean with this specifically:

> When extracting translations we may not be interested about what these custom Twig functions actually do.
But to parse templates, we still want to know which function names are declared.

**Comment by drzraf** [comment]
As long as `init()` isn't call from `__construct()` :) this will be enough.

**Comment by drzraf** [comment]
Oh, I also forgot `public function add_timber_filters( $twig ) {` calling `apply_filters()` at the end of the Twig extension which is also too WordPress-centric during the initial extension of the Twig environment.

**Comment by JUVOJustin** [comment]
So in theory once wp-cli is upgraded to gettext v5 this is resolved and the pot file can be generated? Will it work with `wp-cli timber twig-pot` or with `wp i18n make-pot` since the scanner is baked into gettext?

**Comment by drzraf** [comment]
`wp i18n make-pot` would support Twig out of the box once the scanner is included (in `gettext/gettext` v5).
But as mentioned in https://github.com/wp-cli/i18n-command/issues/303, the issue is that gettext/gettext v5 doesn't support PHP 5.6 that wp-cli/i18n-command wants to keep supporting. As such we are lock.

**Comment by nlemoine** [comment]
Hi everyone,

It has been something missing for a very long time and I'm happy to announce that we finally released an official WP-CLI package that handles WordPress translations extraction for Twig files:
https://github.com/timber/wp-i18n-twig

It overrides the default `i18n make-pot` command, works pretty nice and should cover most Twig use cases (see [current limitations](https://github.com/timber/wp-i18n-twig#limitations) we might tackle in the future).

Any feedback welcome 🙂 

(Closing the ticket since this is now the official way to handle Twig i18n)
